src/main/java/org/sonarsource/scanner/maven/bootstrap/MavenCompilerResolver.java

```diff
@@ -0,0 +1,298 @@
+    PlexusConfiguration pomConfiguration = new XmlPlexusConfiguration(configuration);
+    PlexusConfiguration config = pomConfiguration.getChild(parameterName, false);
+    if (config == null) {
+      return Optional.empty();
+    }
+    return Optional.of(Stream.of(config.getChildren())
+      .collect(Collectors.toMap(PlexusConfiguration::getName, c -> convertString(exec, config.getChild(c.getName(), false)))));
+  }
+
+  private Optional<String> getJdkHome(MojoExecution compilerExecution) {
+
+    Optional<String> executable = getStringConfiguration(compilerExecution, "executable");
+    if (executable.isPresent()) {
+      // Should be something like <jdk home>/bin/javac
+      return Optional.of(Paths.get(executable.get()).getParent().getParent().toString());
+    }
+
+    // Inspired by
+    // https://github.com/apache/maven-compiler-plugin/blob/dc4a5635ba4eb2ba5e461fa53b2c47c58d7fa397/src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java#L1418
+    Toolchain tc = null;
+
+    // Maven 3.3.1 has plugin execution scoped Toolchain Support
+    Optional<Map<String, String>> jdkToolchain = getMapConfiguration(compilerExecution, "jdkToolchain");
+    if (jdkToolchain.isPresent() && !jdkToolchain.get().isEmpty()) {
+      List<Toolchain> tcs = collectMatchingToolchains(jdkToolchain.get());
+      if (!tcs.isEmpty()) {
+        tc = tcs.get(0);
+      }
+    }
+
+    // Fallback on the global jdk toolchain
+    if (tc == null) {
+      tc = toolchainManager.getToolchainFromBuildContext("jdk", session);
+    }
+
+    if (tc instanceof DefaultJavaToolChain) {
+      return Optional.of(((DefaultJavaToolChain) tc).getJavaHome());
+    }
+
+    // Like m-compiler-p, last fallback is to compile with the runtime JDK
+    String runtimeCompiler = getJavacExecutableFromRuntimeJdk();
+    if (runtimeCompiler != null) {
+      return Optional.of(Paths.get(runtimeCompiler).getParent().getParent().toString());
+    }
+
+    return Optional.empty();
+  }
+
+  @SuppressWarnings("unchecked")
+  private List<Toolchain> collectMatchingToolchains(Map<String, String> jdkToolchain) {
+    // getToolchains method only added in Maven 3.3.1, so use reflection
+    try {
+      Method getToolchainsMethod = toolchainManager.getClass().getMethod("getToolchains", MavenSession.class, String.class, Map.class);
+      return (List<Toolchain>) getToolchainsMethod.invoke(toolchainManager, session, "jdk", jdkToolchain);
+    } catch (NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
+      // ignore
+      return Collections.emptyList();
+    }
+  }
+
+  /**
+   * Inspired from https://github.com/codehaus-plexus/plexus-compiler/blob/3300ad47ea45a3c3b9f815acc0973020b21486cd/plexus-compilers/plexus-compiler-javac/src/main/java/org/codehaus/plexus/compiler/javac/JavacCompiler.java#L999
+   */
+  @CheckForNull
+  private static String getJavacExecutableFromRuntimeJdk() {
```

Comment:
Why not using Optional here as well?

Reply:
I initially tried to not change the code too much to allow easier backports of possible future changes on plexus-compiler. But I realize I have already changed the code a lot, so I guess I can use Optional ;)